### DISCUSSION TITLE
Static Maps - saving maps as PNGs

### DISCUSSION THREAD

**Discussion by kyllerss** [protomaps/PMTiles#350] [Q&A]
I was happy to find references to generating static images without violating Term and Conditions restrictions as one of Protomap's value propositions. 

For my own project, I've been looking into how to leverage `PMTiles` + `maplibre-gl` to generate static images of maps. I have a NodeJS app in which I would like to generate map PNGs to store for later use - this is to say that the map generation has to happen in a headless environment. 

[maplibre-native](https://github.com/maplibre/maplibre-native) seems to be the recommended way of doing this w/ maplibre, but I haven't found any information on how to integrate maplibre-native with Protomaps.

Any pointers as to where to find info on generating static images using Protomaps? Thanks!

**Comment by JinIgarashi** [comment]
I implemented static api with pmtiles. See the below implementation 

https://github.com/UNDP-Data/geohub/blob/develop/sites/static-image-api/src/lib/server/renderMap.ts

you can try our api at the below website 
https://staticimage.undpgeohub.org/

or, you maybe can use my maplibre export plugin. This can work without having a server.
https://github.com/watergis/maplibre-gl-export

**Reply by kyllerss** [reply]
Thank you! I'll take a look.

**Reply by bdon** [reply]
MapLibre GL and Mapnik are the only mature options here. The downloads available at https://docs.protomaps.com are only in vector format, though you can use maplibre native to rasterize them; storing raster basemap tiles for the world is impractical however, since typical applications need to zoom to 15+ and at that zoom level it requires storing billions of tiles. Vector tiles can be overzoomed up to ~z20.

**Reply by kyllerss** [reply]
Thanks, @bdon! I hadn't heard of Mapnik. Another option to look at. In my case, I am looking at creating thumbnails for a user feed, so static images is really the only option I have. Thanks for the pointer!

**Comment by do-me** [comment]
@kyllerss 
If you want to generate pngs in the frontend, you can use native maplibre with `map.getCanvas().toDataURL('image/png')`.
In this repo I created e.g. a thumbnail generator (super fast with reduced styling) as well as a full export 1:1 https://github.com/do-me/maplibre-thumbnails

**Comment by bdon** [comment]
addition: I recommend [chiitiler](https://github.com/Kanahiro/chiitiler) as an simple way to run MapLibre Native and generate static images from PMTiles
